linear-algebra/src/main/scala/scalan/linalgebra/Matrices.scala

```diff
@@ -42,46 +54,46 @@ trait Matrices extends Vectors { self: MatricesDsl =>
       DenseVector(mT.rows.map(row => row.nonZeroIndices.length))
     }
 
-    //@OverloadId("vector")
-    def *(vector: Vec[T])(implicit n: Numeric[T]): Vec[T] = {
-      DenseVector(rows.map { r => r.dot(vector) })
-    }
+    def +(matrix: Matr[T])(implicit n: Numeric[T]): Matr[T] = self +^^ matrix
     @OverloadId("matrix")
     def *(matrix: Matr[T])(implicit n: Numeric[T], o: Overloaded1): Matr[T]
-    def +^^(other: Matr[T])(implicit n: Numeric[T]): Matr[T]
+    def *(vector: Vec[T])(implicit n: Numeric[T]): Vec[T] = DenseVector(rows.map(r => r dot vector))
     @OverloadId("matrix")
-    def *^^(other: Matr[T])(implicit n: Numeric[T]): Matr[T]
-    def *^^(value: Rep[T])(implicit n: Numeric[T], o: Overloaded1): Matr[T]
-    def average(implicit f: Fractional[T], m: RepMonoid[T]): DoubleRep
-
-    def companion: Rep[MatrixCompanion]
+    def +^^(matrix: Matr[T])(implicit n: Numeric[T]): Matr[T]
+    def +^^(value: Rep[T])(implicit n: Numeric[T], o: Overloaded1): Matr[T] = {
+      self +^^ ConstMatrix(value, numColumns, numRows)
+    }
+    @OverloadId("matrix")
+    def *^^(matrix: Matr[T])(implicit n: Numeric[T]): Matr[T]
+    def *^^(value: Rep[T])(implicit n: Numeric[T], o: Overloaded1): Matr[T] = {
+      self +^^ ConstMatrix(value, numColumns, numRows)
+    }
+    def average(implicit f: Fractional[T], m: RepMonoid[T]): Rep[T]
   }
 
   abstract class DenseFlatMatrix[T](val rmValues: Rep[Collection[T]], val numColumns: Rep[Int])
                                    (implicit val eT: Elem[T]) extends Matrix[T] {
 
     def items = rmValues
-    def companion = DenseFlatMatrix
+    def constItem = zeroValue
```

Comment:
I didn't see usage for `companion` method, and had an idea to put `constItem` into `Matrix` trait interface. But abandoned it and removed (just not ready to PR those changes).